### PR TITLE
GH-857: Memory management and interfaces

### PR DESCRIPTION
**Pull request by alanakbik**
This PR introduces a number of changes to address memory management and slim down some interfaces. 

For **memory management**, it introduces the `embedding_storage_mode` parameter to the `ModelTrainer` class and `predict()` methods. This parameter can be one of 'none', 'cpu' and 'gpu':
- If 'none' all embeddings are deleted after usage - this has lowest memory requirements but means that embeddings need to be recomputed at each epoch of training potentially causing a slowdown. 
- If 'cpu' all embeddings are moved to CPU memory after usage. During training, this means that they only need to be moved back to GPU for the forward pass, and not recomputed so in many cases this is faster, but requires memory. 
- If 'gpu' all embeddings stay on GPU memory after computation. This eliminates memory shuffling during training, causing a speedup. However this option requires enough GPU memory to be available for all embeddings of the dataset. 

To use this option during training, simply set the parameter: 

```python
        # initialize trainer
        trainer: ModelTrainer = ModelTrainer(tagger, corpus)
        trainer.train(
            "path/to/your/model",
            embedding_storage_mode='gpu',
        )
```

In terms of **interfaces**, this PR introduces the `DataPoint` interface that is currently implemented by the `Token` and `Sentence` classes. The interface currently only has a `embedding` property, meaning that each `DataPoint` in Flair needs to be able to get an embedding. We will utilize this interface in the near future to extend Flair to other data types besides words and sentences.

It also slims down the `evaluate()` method in the `flair.nn.Model` interface to take a `DataLoader` instead of a group of parameters.

### CODE REVIEW

**Comment by yosipk** [comment]
👍

**Comment by alanakbik** [comment]
:+1:
